Sources/Kafka/KafkaConsumer.swift

```diff
@@ -331,26 +352,89 @@ public final class KafkaConsumer: Sendable, Service {
+            switch nextAction {
+            case .pollForEvents(let client):
+                // Event poll to serve any events queued inside of `librdkafka`.
+                _ = client.eventPoll()
+                try await Task.sleep(for: self.configuration.pollInterval)
+            case .terminatePollLoop:
+                return
+            }
+        }
+    }
+
+    /// Run loop polling Kafka for new consumer messages.
+    private func messageRunLoop() async throws {
         while !Task.isCancelled {
-            let nextAction = self.stateMachine.withLockedValue { $0.nextPollLoopAction() }
+            let nextAction = self.stateMachine.withLockedValue { $0.nextConsumerPollLoopAction() }
             switch nextAction {
             case .pollForAndYieldMessage(let client, let source):
-                let events = client.eventPoll()
-                for event in events {
-                    switch event {
-                    case .consumerMessages(let result):
-                        // We do not support back pressure, we can ignore the yield result
-                        _ = source.yield(result)
-                    default:
-                        break // Ignore
+                // Poll for new consumer message.
+                var result: Result<KafkaConsumerMessage, Error>?
+                do {
+                    if let message = try client.consumerPoll() {
+                        result = .success(message)
+                    }
+                } catch {
+                    result = .failure(error)
+                }
+                if let result {
+                    let yieldResult = source.yield(result)
+                    switch yieldResult {
+                    case .produceMore:
+                        break
+                    case .stopProducing:
+                        self.stateMachine.withLockedValue { $0.stopProducing() }
+                    case .dropped:
+                        break
+                    }
+                } else {
+                    self.stateMachine.withLockedValue { $0.waitForNewMessages() }
+                }
+            case .pollForMessageAndSleep(let client, let source):
+                var result: Result<KafkaConsumerMessage, Error>?
+                do {
+                    if let message = try client.consumerPoll() {
+                        result = .success(message)
                     }
+                } catch {
+                    result = .failure(error)
+                }
+                if let result {
+                    self.stateMachine.withLockedValue { $0.produceMore() }
+                    let yieldResult = source.yield(result)
+                    switch yieldResult {
+                    case .produceMore:
+                        break
+                    case .stopProducing:
+                        self.stateMachine.withLockedValue { $0.stopProducing() }
```

Comment:
It seems If we suspend producing here, we should not reach `newMessagesProduced`, otherwise we bump into `fatalError`:
https://github.com/swift-server/swift-kafka-client/pull/139/files/6686e982c3c1c905acc0a6f4653cdf903f6738ff#diff-8ba8c17c39d40fc29ec22c104d01bacc928df36e717000547f2a97d4203ef6a1R797

Reply:
Yes indeed, very good catch!